File: api/src/main/java/io/xol/chunkstories/api/net/packets/PacketDecal.java
package io.xol.chunkstories.api.net.packets;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import io.xol.chunkstories.api.client.net.ClientPacketsProcessor;
import io.xol.chunkstories.api.exceptions.PacketProcessingException;
import io.xol.chunkstories.api.math.vector.dp.Vector3dm;
import io.xol.chunkstories.api.net.PacketDestinator;
import io.xol.chunkstories.api.net.PacketSynchPrepared;
import io.xol.chunkstories.api.net.PacketsProcessor;
import io.xol.chunkstories.api.net.PacketSender;

//(c) 2015-2017 XolioWare Interactive
//http://chunkstories.xyz
//http://xol.io

/** Simply sends a decal to the client to be drawn */
public class PacketDecal extends PacketSynchPrepared
{
	public String decalName;
	public Vector3dm position;
	public Vector3dm orientation;
	public Vector3dm size;

	@Override
	public void sendIntoBuffer(PacketDestinator destinator, DataOutputStream out) throws IOException
	{
		out.writeUTF(decalName);
		out.writeDouble(position.getX());
		out.writeDouble(position.getY());
		out.writeDouble(position.getZ());
		out.writeDouble(orientation.getX());
		out.writeDouble(orientation.getY());
		out.writeDouble(orientation.getZ());
		out.writeDouble(size.getX());
		out.writeDouble(size.getY());
		out.writeDouble(size.getZ());
	}

	@Override
	public void process(PacketSender sender, DataInputStream in, PacketsProcessor processor) throws IOException, PacketProcessingException
	{
		decalName = in.readUTF();
		position = new Vector3dm();
		position.setX(in.readDouble());
		position.setY(in.readDouble());
		position.setZ(in.readDouble());

		orientation = new Vector3dm();
		orientation.setX(in.readDouble());
		orientation.setY(in.readDouble());
		orientation.setZ(in.readDouble());

		size = new Vector3dm();
		size.setX(in.readDouble());
		size.setY(in.readDouble());
		size.setZ(in.readDouble());
		
		if(processor instanceof ClientPacketsProcessor)
		{
			ClientPacketsProcessor cpp = (ClientPacketsProcessor)processor;
			cpp.getContext().getDecalsManager().drawDecal(position, orientation, size, decalName);
		}
	}

}

File: common/src/main/java/xyz/chunkstories/net/packets/PacketDecal.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//
package xyz.chunkstories.net.packets

import org.joml.Vector3d
import org.joml.Vector3dc
import xyz.chunkstories.api.net.PacketWorld
import xyz.chunkstories.api.player.Player
import xyz.chunkstories.api.world.World
import xyz.chunkstories.api.world.WorldClient
import java.io.DataInputStream
import java.io.DataOutputStream

/** Simply sends a decal to the client to be drawn  */
class PacketDecal : PacketWorld {
    private lateinit var decalName: String
    private lateinit var position: Vector3dc
    private lateinit var orientation: Vector3dc
    private lateinit var size: Vector3dc

    constructor(world: World) : super(world)
    constructor(world: World, decalName: String, position: Vector3dc, orientation: Vector3dc, size: Vector3dc) : super(world) {
        this.decalName = decalName
        this.position = position
        this.orientation = orientation
        this.size = size
    }

    override fun send(dos: DataOutputStream) {
        dos.writeUTF(decalName)
        dos.writeDouble(position.x())
        dos.writeDouble(position.y())
        dos.writeDouble(position.z())
        dos.writeDouble(orientation.x())
        dos.writeDouble(orientation.y())
        dos.writeDouble(orientation.z())
        dos.writeDouble(size.x())
        dos.writeDouble(size.y())
        dos.writeDouble(size.z())
    }

    override fun receive(dis: DataInputStream, player: Player?) {
        decalName = dis.readUTF()
        val position = Vector3d()
        position.x = dis.readDouble()
        position.y = dis.readDouble()
        position.z = dis.readDouble()
        val orientation = Vector3d()
        orientation.x = dis.readDouble()
        orientation.y = dis.readDouble()
        orientation.z = dis.readDouble()
        val size = Vector3d()
        size.x = dis.readDouble()
        size.y = dis.readDouble()
        size.z = dis.readDouble()
        if (world is WorldClient) {
            world.decalsManager.add(position, orientation, size, decalName)
        }
    }
}